source/adapters/level_zero/command_buffer.cpp

```diff
@@ -1168,67 +1373,127 @@ ur_result_t waitForDependencies(ur_exp_command_buffer_handle_t CommandBuffer,
 }
 
 /**
- * Creates a host visible event and appends a barrier to signal it when the
- * command buffer finishes executing.
- * @param[in] CommandBuffer The command buffer.
- * @param[in] Queue The UR queue used to submit the command buffer.
- * @param[in] SignalCommandList The command-list to append the barrier to.
- * @param[out][optional] Event The host visible event which will be returned
- * to the user, if user passed an output parameter to the UR API.
- * @return UR_RESULT_SUCCESS or an error code on failure
+ * Appends a QueryKernelTimestamps command that does profiling for all the
+ * sync-point events.
+ * @param CommandBuffer The command-buffer that is being enqueued.
+ * @param CommandList The command-list to append the QueryKernelTimestamps
+ * command to.
+ * @param SignalEvent The event that must be signaled after the profiling is
+ * finished. This event will contain the profiling information.
+ * @param WaitEvent The event that must be waited on before starting the
+ * profiling.
+ * @return UR_RESULT_SUCCESS or an error code on failure.
  */
-ur_result_t createUserEvent(ur_exp_command_buffer_handle_t CommandBuffer,
-                            ur_queue_handle_t Queue,
-                            ur_command_list_ptr_t SignalCommandList,
-                            ur_event_handle_t *Event) {
-  // Execution event for this enqueue of the UR command-buffer
-  ur_event_handle_t RetEvent{};
-
-  UR_CALL(createEventAndAssociateQueue(Queue, &RetEvent,
-                                       UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP,
-                                       SignalCommandList, false, false, true));
-
-  if ((Queue->Properties & UR_QUEUE_FLAG_PROFILING_ENABLE) &&
-      (!CommandBuffer->IsInOrderCmdList) &&
-      (CommandBuffer->IsProfilingEnabled)) {
-    // Multiple submissions of a command buffer implies that we need to save
-    // the event timestamps before resubmiting the command buffer. We
-    // therefore copy the these timestamps in a dedicated USM memory section
-    // before completing the command buffer execution, and then attach this
-    // memory to the event returned to users to allow to allow the profiling
-    // engine to recover these timestamps.
-    command_buffer_profiling_t *Profiling = new command_buffer_profiling_t();
-
-    Profiling->NumEvents = CommandBuffer->ZeEventsList.size();
-    Profiling->Timestamps =
-        new ze_kernel_timestamp_result_t[Profiling->NumEvents];
-
-    ZE2UR_CALL(zeCommandListAppendQueryKernelTimestamps,
-               (SignalCommandList->first, CommandBuffer->ZeEventsList.size(),
-                CommandBuffer->ZeEventsList.data(),
-                (void *)Profiling->Timestamps, 0, RetEvent->ZeEvent, 1,
-                &(CommandBuffer->SignalEvent->ZeEvent)));
-
-    RetEvent->CommandData = static_cast<void *>(Profiling);
-  } else {
+ur_result_t doProfiling(ur_exp_command_buffer_handle_t CommandBuffer,
```

Comment:
Nit but `doProfiling` is not a very descriptive function name to me, would prefer something more like `appendProfilingQueries` or something like that.